src/Elliptic/SubdomainPreconditioners/MinusLaplacian.hpp

```diff
@@ -0,0 +1,261 @@
+ */
+template <size_t Dim, typename OptionsGroup,
+          typename Solver = LinearSolver::Serial::LinearSolver<tmpl::list<
+              ::LinearSolver::Serial::Registrars::Gmres<
+                  ::LinearSolver::Schwarz::ElementCenteredSubdomainData<
+                      Dim, tmpl::list<Poisson::Tags::Field>>>,
+              ::LinearSolver::Serial::Registrars::ExplicitInverse>>,
+          typename LinearSolverRegistrars =
+              tmpl::list<Registrars::MinusLaplacian<Dim, OptionsGroup, Solver>>>
+class MinusLaplacian
+    : public LinearSolver::Serial::LinearSolver<LinearSolverRegistrars> {
+ private:
+  using Base = LinearSolver::Serial::LinearSolver<LinearSolverRegistrars>;
+  using StoredSolverType = tmpl::conditional_t<std::is_abstract_v<Solver>,
+                                               std::unique_ptr<Solver>, Solver>;
+
+ public:
+  static constexpr size_t volume_dim = Dim;
+  using options_group = OptionsGroup;
+  using poisson_system =
+      Poisson::FirstOrderSystem<Dim, Poisson::Geometry::FlatCartesian>;
+  using SubdomainOperator =
+      elliptic::dg::subdomain_operator::SubdomainOperator<poisson_system,
+                                                          OptionsGroup>;
+  using SubdomainData = ::LinearSolver::Schwarz::ElementCenteredSubdomainData<
+      Dim, tmpl::list<Poisson::Tags::Field>>;
+  using solver_type = Solver;
+
+  struct SolverOptionTag {
+    static std::string name() noexcept { return "Solver"; }
+    using type = StoredSolverType;
+    static constexpr Options::String help =
+        "The linear solver used to invert the Laplace operator. The solver is "
+        "shared between the tensor components.";
+  };
+
+  using options = tmpl::list<SolverOptionTag>;
+  static constexpr Options::String help =
+      "Approximate the linear operator with a Laplace operator with Dirichlet "
+      "boundary conditions for every tensor component separately.";
+
+  MinusLaplacian() = default;
+  MinusLaplacian(MinusLaplacian&& /*rhs*/) = default;
+  MinusLaplacian& operator=(MinusLaplacian&& /*rhs*/) = default;
+  ~MinusLaplacian() = default;
+  MinusLaplacian(const MinusLaplacian& rhs) noexcept
+      : Base(rhs), solver_(rhs.clone_solver()) {}
+  MinusLaplacian& operator=(const MinusLaplacian& rhs) noexcept {
+    Base::operator=(rhs);
+    solver_ = rhs.clone_solver();
+    return *this;
+  }
+
+  /// \cond
+  explicit MinusLaplacian(CkMigrateMessage* m) noexcept : Base(m) {}
+  using PUP::able::register_constructor;
+  WRAPPED_PUPable_decl_template(MinusLaplacian);  // NOLINT
+  /// \endcond
+
+  MinusLaplacian(StoredSolverType solver) noexcept
+      : solver_(std::move(solver)) {}
+
+  const Solver& solver() const noexcept {
+    if constexpr (std::is_abstract_v<Solver>) {
+      return *solver_;
+    } else {
+      return solver_;
+    }
+  }
+  Solver& solver() noexcept {
+    if constexpr (std::is_abstract_v<Solver>) {
+      return *solver_;
```

Comment:
How so? They just return a reference to the object held by the unique_ptr. What am I missing?

Reply:
as far as I understand, I can call this function and get the reference. Then the MinusLaplacian can get destructed which would destruct the `solver_` object because the unique pointer has no way of accounting for any references to the objects that may still exist. The my reference is dangling, or not?

Reply:
Makes sense. Another basic C++ question then: How is this avoided for accessors that return a const-ref? I haven't seen accessor functions account for this sort of thing yet.

Reply:
Isn't it just a programmer error to hold refs into an object that gets destructed?

Reply:
I am not sure. I think usually it is left to the user to make sure that all references they hold are referencing objects that still exist.

Reply:
https://blog.galowicz.de/2016/03/23/const_reference_to_temporary_object/
Ok I learned something today. So I just make the non-const accessor private?

Reply:
I think this article is about const rvalue references which is not the case here.

Reply:
`shared_ptr` is discouraged in SpECTRE, mostly because there is no reasonable way to serialize a multiply owned object.

> I think this article is about const rvalue references which is not the case here.

Const references, I think of either type, to temporaries of various sorts.  But yes, not relevant.

> I think usually it is left to the user to make sure that all references they hold are referencing objects that still exist.

Yes.  C++ is not designed to be inherently memory safe.  As C++ programmers we have to accept that we must manually manage lifetimes.

> But in this case I do not like that you are taking a reference of a `unique_ptr` object, that sort of defeats the purpose (other than automatic destruction).

Not sure what your objection is here.  `unique_ptr<X>` is conceptually the same as a variable of type `X`, but on the heap rather than the stack.  Both have an "ownership" quality, which is the opposite of `X&`, which is access without ownership.  A reference to a `unique_ptr` object itself would be kind of odd, but that's not what is happening here.

Reply:
alright, I guess that is fair, then leave it as it is. Can you delete the non-const ref accessor though once/if the buffers of the solver are mutable and the call operators are `const`?

Reply:
I made the non-const accessor private now. It's needed for the `reset` function.